File: src/main/java/hzt/aoc/day03/TreesEncounteredPart2.java
package hzt.aoc.day03;

import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;

import java.awt.*;
import java.util.List;

public class TreesEncounteredPart2 extends Day03Challenge {


    private static final Logger LOGGER = LogManager.getLogger(TreesEncounteredPart2.class);


    public TreesEncounteredPart2() {
        super("part 2",
                "Find the product of the number of trees crossed by all the given paths");
    }

    @Override
    protected long calculateResult(List<List<Boolean>> grid) {
        long product = 1;
        for (Path path : Path.values()) {
            int numberOfTrees = calculateNumberOfTreesEncountered(grid, new Point(0, 0), path.getSlope());
            LOGGER.info(String.format("The number of trees crossed using %s is %d", path.name(), numberOfTrees));
            product *= numberOfTrees;
        }
        LOGGER.info("");
        return product;
    }

    @Override
    public String getMessage(long result) {
        return String.format("The product of all the number of trees crossed is: %d%n", result);
    }

}

File: src/main/kotlin/hzt/aoc/day03/TreesEncounteredPart2.kt
package hzt.aoc.day03

import java.awt.Point

class TreesEncounteredPart2 : Day03Challenge(
    "part 2",
    "Find the product of the number of trees crossed by all the given paths"
) {
    override fun calculateResult(grid: List<List<Boolean>>): Long {
        var product: Long = 1
        for (path in Path.values()) {
            val numberOfTrees = calculateNumberOfTreesEncountered(grid, Point(0, 0), path.slope)
            LOGGER.info(
                String.format(
                    "The number of trees crossed using %s is %d",
                    path.name,
                    numberOfTrees
                )
            )
            product *= numberOfTrees.toLong()
        }
        LOGGER.info("")
        return product
    }

    override fun getMessage(result: String): String {
        return String.format("The product of all the number of trees crossed is: %s%n", result)
    }
}
